x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/changepoint/ChangePointAggregator.java

```diff
@@ -279,22 +231,289 @@ static double[] outlierWeights(double[] values) {
+            double ru = upperLeastSquares.rSquared();
+            double vl = lowerRange.variance() * (1.0 - rl);
+            double vu = upperRange.variance() * (1.0 - ru);
+            double v = (nl * vl + nu * vu) / (nl + nu);
+            if (v < vChange) {
+                vChange = v;
+                changePoint = cp;
+            }
+        }
+
+        double pValue = independentTrialsPValue(
+            fTestNestedPValue(H0.dataStats().nValues(), H0.var(), H0.nParams(), vChange, 6.0),
+            candidateChangePoints.length
+        );
+
+        return new TestStats(Type.TREND_CHANGE, pValue, vChange, 6.0, changePoint, H0.dataStats());
+    }
+
+    static TestStats testVsStepChange(
+        TestStats trendChange,
+        double[] values,
+        double[] weights,
+        int[] candidateChangePoints,
+        double pValueThreshold
+    ) {
+        DataStats dataStats = trendChange.dataStats();
+        TestStats stationary = new TestStats(Type.STATIONARY, 1.0, dataStats.var(), 1.0, dataStats);
+        TestStats stepChange = testStepChangeVs(stationary, values, weights, candidateChangePoints);
+        double n = dataStats.nValues();
+        double pValue = fTestNestedPValue(n, stepChange.var(), 2.0, trendChange.var(), 6.0);
+        return pValue < pValueThreshold ? trendChange : stepChange;
+    }
+
+    static double fTestNestedPValue(double n, double vNull, double pNull, double vAlt, double pAlt) {
+        if (vAlt == vNull) {
             return 1.0;
         }
-        if (varianceAlt == 0.0) {
+        if (vAlt == 0.0) {
             return 0.0;
         }
-        double F = dfAlt / dfNull * vNull / varianceAlt;
-        double sf = fDistribSf(dfNull, dfAlt, F);
+        double F = (vNull - vAlt) / (pAlt - pNull) * (n - pAlt) / vAlt;
+        double sf = fDistribSf(pAlt - pNull, n - pAlt, F);
         return Math.min(2 * sf, 1.0);
     }
 
+    static Tuple<double[], double[]> sample(double[] values, double[] weights) {
+        if (values.length <= 500) {
+            return Tuple.tuple(values, weights);
+        }
+
+        // Just want repeatable random numbers.
+        Random rng = new Random(126832678);
+        UniformRealDistribution uniform = new UniformRealDistribution(RandomGeneratorFactory.createRandomGenerator(rng), 0.0, 0.99999);
+
+        // Fisher–Yates shuffle (why isn't this in Arrays?).
+        int[] choice = IntStream.range(0, values.length).toArray();
+        for (int i = 0; i < 500; ++i) {
+            int index = i + (int) Math.floor(uniform.sample() * (values.length - i));
+            int tmp = choice[i];
+            choice[i] = choice[index];
+            choice[index] = tmp;
+        }
+
+        double[] sample = new double[500];
+        double[] sampleWeights = new double[500];
+        Arrays.sort(choice, 0, 500);
+        for (int i = 0; i < 500; ++i) {
+            sample[i] = values[choice[i]];
+            sampleWeights[i] = weights[choice[i]];
+        }
```

Comment:
Unfortunately, the swap in line 384 requires that you be able to copy into the full array (using arrays).

One strategy for doing this without creating the array is to use a hash map and deduce that any miss is equal to the request index. I don't know the trade offs as well for Java, but for definite that approach would be slower than using the array approach in C/C++ given the likely max bucket count we have to deal with.

So to be clear I would have to create a reserved hash map (of size equal to the sample size) of int -> int which I use to keep track of indices I've swapped out of the first sample size. Does this sound like it would be better for Java? I guess I could time it.

Reply:
With JDK21, I am just being sensitive to creating too much garbage. A hashmap probably would create as much or more garbage and possibly a even larger single object. `500` integers is only 2000 bytes, so we are probably ok.

Reply:
Yes. It is a balance for definite. I wondered if you were also worried about the memory pressure aspect. I think for realistic uses the array will probably be better on memory pressure as well. One should probably use a cut over in strategy for very high count, but that does seem a bit like overkill.